Data Analysis

Importing the necessary libraries

In [1]:
import pandas as pd
import numpy as np
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


Loading the data set into pandas dataframe

In [2]:
data = pd.read_csv("/content/Data Analyst - Test Data - US.csv")

In [3]:
# first five rows
data.head()

Review        date     Location
0  I was very impressed with the resort.\n Great ...  2019/08/20    Sebastian
1  The rooms were nice the outside needs work als...  2019/08/20  Los Angeles
2  Great location! I have stayed at this hotel on...  2019/08/20      Georgia
3  The hotel was adequate for my stay. The strips...  2019/08/20          NaN
4  Great location, room was large and spacious. P...  2019/08/19  Palm Harbor

In [4]:
# Rows and columns of the data set
data.shape

(6448, 3)

In [5]:
# checkiing the data types 
data.dtypes

Review      object
date        object
Location    object
dtype: object

All the columns are of object data type

Checking the missing values in the data set

In [6]:
data.isnull().sum()

Review        55
date           0
Location    4737
dtype: int64

Review column has 55 null values and Location has 4737 null values

Dropping all the null values from the columns

In [7]:
data.dropna(inplace=True)

In [8]:
data.isnull().sum()

Review      0
date        0
Location    0
dtype: int64

All the null values are dropped

Dropping Date and Location columns

In [11]:
data.drop(columns=["date","Location"],axis=1,inplace=True)

In [12]:
data.head()

Review
0   I was very impressed with the resort.\n Great ...
1   The rooms were nice the outside needs work als...
2   Great location! I have stayed at this hotel on...
4   Great location, room was large and spacious. P...
10  Very clean and friendly and I love the locatio...

Converting from object to string data type

In [15]:
data["Review"]=data['Review'].astype(str)

word_count : how many words are there in each Review

In [16]:
data["word_count"]=data["Review"].apply(lambda x: len(x.split()))

In [17]:
data.head()

Review  word_count
0   I was very impressed with the resort.\n Great ...          33
1   The rooms were nice the outside needs work als...          25
2   Great location! I have stayed at this hotel on...          20
4   Great location, room was large and spacious. P...          20
10  Very clean and friendly and I love the locatio...          40

character count :how many characters are there in each review

In [18]:
data["char_count"]=data["Review"].apply(lambda x:len(x))

In [19]:
data.head()

Review  word_count  char_count
0   I was very impressed with the resort.\n Great ...          33         196
1   The rooms were nice the outside needs work als...          25         134
2   Great location! I have stayed at this hotel on...          20         106
4   Great location, room was large and spacious. P...          20         126
10  Very clean and friendly and I love the locatio...          40         238

Finding the average words in each of the Review

In [20]:
def average_words(x):      # function to find the avarage words
  words=x.split()
  return sum(len(word) for word in words)/len(words)

In [21]:
data["average_wordcount"]=data["Review"].apply(lambda x:average_words(x))

In [22]:
data.head()

Review  ...  average_wordcount
0   I was very impressed with the resort.\n Great ...  ...            4.69697
1   The rooms were nice the outside needs work als...  ...            4.12000
2   Great location! I have stayed at this hotel on...  ...            4.00000
4   Great location, room was large and spacious. P...  ...            5.00000
10  Very clean and friendly and I love the locatio...  ...            4.80000

[5 rows x 4 columns]

Finding out the stop words 

In [23]:
stop_words=stopwords.words("english")
len(stop_words)

179

In [24]:
data["stopword_count"] = data['Review'].apply(lambda x:len([word for word in x.split() if word.lower() in stop_words]))

In [25]:
data.head()

Review  ...  stopword_count
0   I was very impressed with the resort.\n Great ...  ...              14
1   The rooms were nice the outside needs work als...  ...              10
2   Great location! I have stayed at this hotel on...  ...               8
4   Great location, room was large and spacious. P...  ...               7
10  Very clean and friendly and I love the locatio...  ...              19

[5 rows x 5 columns]

stopword_rate:percentage of stop word to non stop word

In [27]:
data['stopword_rate'] = data["stopword_count"]/data["word_count"]

In [28]:
data.head()

Review  ...  stopword_rate
0   I was very impressed with the resort.\n Great ...  ...       0.424242
1   The rooms were nice the outside needs work als...  ...       0.400000
2   Great location! I have stayed at this hotel on...  ...       0.400000
4   Great location, room was large and spacious. P...  ...       0.350000
10  Very clean and friendly and I love the locatio...  ...       0.475000

[5 rows x 6 columns]

Reviews with high stop words

In [30]:
data.sort_values(by="stopword_rate",ascending=False)

Review  ...  stopword_rate
6267  ·it was so close to what we were there for!\n ...  ...       0.636364
1084  I like the location. It is close to the air po...  ...       0.593750
4553  property was partially under construction duri...  ...       0.588235
2404  The only thing we liked about the property was...  ...       0.588235
3517  There was no elevator, I have a mother who is ...  ...       0.587302
...                                                 ...  ...            ...
6345  ·housekeeping staff\n \n \n \n \n ·restaurant com  ...       0.000000
6365                                           ·the bed  ...       0.000000
6379                         ·staff, cleanliness, price  ...       0.000000
6385                                      ·no microwave  ...       0.000000
6333          ·great rooms, great service, great value.  ...       0.000000

[1705 rows x 6 columns]

summary of the data set

In [31]:
data.describe()

word_count   char_count  ...  stopword_count  stopword_rate
count  1705.000000  1705.000000  ...     1705.000000    1705.000000
mean     44.531378   250.558944  ...       18.478592       0.369202
std      44.401263   234.884032  ...       22.331574       0.131761
min       1.000000     8.000000  ...        0.000000       0.000000
25%      18.000000   106.000000  ...        5.000000       0.300000
50%      28.000000   165.000000  ...       11.000000       0.400000
75%      55.000000   303.000000  ...       24.000000       0.466667
max     311.000000  1531.000000  ...      174.000000       0.636364

[8 rows x 5 columns]